iroh/src/endpoint.rs

```diff
@@ -587,24 +587,32 @@ impl Endpoint {
     /// endpoint must support this `alpn`, otherwise the connection attempt will fail with
     /// an error.
     pub async fn connect(&self, node_addr: impl Into<NodeAddr>, alpn: &[u8]) -> Result<Connection> {
-        self.connect_with(node_addr, alpn, self.static_config.transport_config.clone())
-            .await
-    }
-
-    /// Connects to a remote [`Endpoint`] using a custom [`TransportConfig`].
-    ///
-    /// Like [`Endpoint::connect`], but allows providing a custom for the connection.  See
-    /// the docs of [`Endpoint::connect`] for details.
-    ///
-    /// Please be aware that changing some settings may have adverse effects on establishing
-    /// and maintaining direct connections.  Carefully test settings you use and consider
-    /// this currently as still rather experimental.
-    pub async fn connect_with(
+        let connecting = self
+            .connect_with_opts(node_addr, alpn, Default::default())
+            .await?;
+        Ok(connecting.await?)
+    }
+
+    /// Starts a connection attempt with a remote [`Endpoint`].
+    ///
+    /// Like [`Endpoint::connect`] (see also its docs for general details), but allows for a more
+    /// advanced connection setup with more customization in two aspects:
+    /// 1. The returned future resolves to a [`Connecting`], which can be further processed into
+    ///    a [`Connection`] by awaiting, or alternatively allows connecting with 0RTT via
+    ///    [`Connecting::into_0rtt`].
+    ///    **Note:** Please read the documentation for `into_0rtt` carefully to assess
+    ///    security concerns.
+    /// 2. The [`TransportConfig`] for the connection can be modified via the provided
+    ///    [`ConnectOptions`].
+    ///    **Note:** Please be aware that changing transport config settings may have adverse effects on
+    ///    establishing and maintaining direct connections.  Carefully test settings you use and
+    ///    consider this currently as still rather experimental.
+    pub async fn connect_with_opts(
         &self,
         node_addr: impl Into<NodeAddr>,
         alpn: &[u8],
-        transport_config: Arc<TransportConfig>,
-    ) -> Result<Connection> {
+        options: ConnectOptions,
```

Comment:
I generally like this approach and will attempt to push this one notch further: There's no need to **require** the ALPN.  If you don't use the router it is perfectly possible to accept a connection without an ALPN.  Why not put the ALPN into the `ConnectOptions` as an Option as well?

Reply:
Hm. Interesting idea. I thought we'd aligned on "iroh connections will always have an ALPN" somewhere in discord though. I think it might be a good nudge for users to always implement signaling on *some* level.

Reply:
Ok, totally fair.  Let's not worry about ALPNs now and leave it as this.